### ISSUE TITLE
[fvt] rflash does not work against 8335-GTA physical machine

### ISSUE THREAD

**Issue by hu-weihua** [xcat2/xcat-core#3972]
XCAT version: 2.13.7 (git commit df8c223afba0daaa53bcbcaf571a8fbc5b823fbf, built Tue Sep 19 07:15:37 EDT 2017)

``rflash`` has problem to resolve command line and to do upgrade job against 8``335-GTA`` physical machine

```
[root@c910f03c05k25 firmware]# ll
total 688260
-rw-r--r-- 1 root root 100663905 Mar 27 13:34 8335GTB_820.1642.20170327n.hpm
-rw-r--r-- 1 root root 100663905 Mar 27 13:34 8335GTB_820.1642.20170327n_pass1.hpm
-rw-r--r-- 1 root root  67109473 Mar 27 13:34 8335GTB_820.1642.20170327n_update.hpm
-rw-r--r-- 1 root root  67109473 Mar 27 13:34 8335GTB_820.1642.20170327n_update_pass1.hpm
-rw-r--r-- 1 root root 100663905 Mar 27 12:58 8335_820.1642.20170327n.hpm
-rw-r--r-- 1 root root    110137 Sep 20 04:33 8335_820.1642.20170327n.html
-rw-r--r-- 1 root root  67109473 Mar 27 12:58 8335_820.1642.20170327n_update.hpm
......

[root@c910f03c05k25 firmware]# pwd
/tmp/firmware

[root@c910f03c05k25 firmware]# rflash fs4 8335_820.1642.20170327n_update.hpm
Error: fs4: Data directory must be specified.

[root@c910f03c05k25 firmware]# rflash fs4 /tmp/firmware/8335_820.1642.20170327n_update.hpm
Error: fs4: Data directory must be specified.

[root@c910f03c05k25 firmware]#   rflash fs4 -d /tmp/firmware/
Error: fs4: The image file doesn't match this machine

[root@c910f03c05k25 firmware]# rflash fs4 /tmp/firmware/8335_820.1642.20170327n_update.hpm -d /tmp/firmware
Error: fs4: Unable to determine firmware version currently installed. Verify that "/opt/xcat/bin/ipmitool-xcat -H 50.3.25.1 -I lanplus -U ADMIN -P admin fru print 47 | grep OP8_v" command returns a version.

[root@c910f03c05k25 firmware]# /opt/xcat/bin/ipmitool-xcat -H 50.3.25.1 -I lanplus -U ADMIN -P admin fru print 47
 Product Name          : OpenPOWER Firmware
 Product Version       : IBM-firestone-ibm-OP8_PFD_v1.6_0.11
 Product Extra         : 	hostboot-1ed7acd-7d144cb
 Product Extra         : 	occ-9427606
 Product Extra         : 	skiboot-5.1.1-7bf1897
 Product Extra         : 	hostboot-binaries-43d5a59
 Product Extra         : 	firestone-xml-9163bac
 Product Extra         : 	capp-ucode-105cb8f

[root@c910f03c05k25 firmware]# /opt/xcat/bin/ipmitool-xcat -H 50.3.25.1 -I lanplus -U ADMIN -P admin fru print 47 | grep OP8_v
[root@c910f03c05k25 firmware]#

```

**Comment by whowutwut** [comment]
I think this issue requires some refactoring of the ipmi.pm code to test the following: 
```
if (openPower) { 
   if (S822LC for HPC) {  
   } else if (S822LC for Big Data) { 
   } else {
      rflash is not supported for this machine
  }
}
```

the way it's written today means that we are going into code path that is not intended.  We also saw this issue on flashing Boston server that had a MTM that was not tested by xCAT, but in fact it fell into the OpenPower code and printed a message from there.

**Comment by gurevichmark** [comment]
Opened a separate issue #3975 to address refactoring ipmi.pm

**Comment by zet809** [comment]
The fix in #3973 will be included in 2.13.7, but let move this issue to 2.13.8 to refine the code logic of ipmi.pm to deal with rflash. Thx!

**Comment by hu-weihua** [comment]
The new test result of fix #3973 
```
[root@c910f02c05p03 tmp]# rinv  fs3 firm
fs3: BMC Firmware: 2.13
fs3: System Firmware Product Version: IBM-firestone-ibm-OP8_v1.7_1.62
fs3: System Firmware Product Additional Info: hostboot-bc98d0b-1a29dff
fs3: System Firmware Product Additional Info: occ-0362706-16fdfa7
fs3: System Firmware Product Additional Info: skiboot-5.1.13
fs3: System Firmware Product Additional Info: hostboot-binaries-43d5a59
fs3: System Firmware Product Additional Info: firestone-xml-e7b4fa2-c302f0e
fs3: System Firmware Product Additional Info: capp-ucode-105cb8f

[root@c910f02c05p03 firmware]# pwd
/tmp/firmware

[root@c910f02c05p03 firmware]# rflash fs3 8335_820.1642.20170327n_update.hpm
fs3: rflash started, upgrade failure will be retried up to 2 times. Please wait...
fs3: Success updating firmware.

[root@c910f02c05p03 firmware]#  rinv fs3 firm
fs3: BMC Firmware: 2.13
fs3: System Firmware Product Version: IBM-firestone-ibm-OP8_v1.12_2.70
fs3: System Firmware Product Additional Info: op-build-0a38e56
fs3: System Firmware Product Additional Info: buildroot-211bd05
fs3: System Firmware Product Additional Info: skiboot-5.4.3
fs3: System Firmware Product Additional Info: hostboot-2eb7706-837c96a
fs3: System Firmware Product Additional Info: linux-4.4.30-openpower1-a5bfb1d
fs3: System Firmware Product Additional Info: petitboot-v1.3.2-9a212b5
fs3: System Firmware Product Additional Info: firestone-xml-2494a43
fs3: System Firmware Product Additional Info: occ-d7efe30-d919b07
fs3: System Firmware Product Additional Info: hos

[root@c910f02c05p03 firmware]# rflash fs3 -d /tmp/firmware/
Error: fs3: The image file doesn't match this machine

[root@c910f02c05p03 firmware]# rflash fs3 -d /tmp/firmware/8335_820.1642.20170327n_update.hpm
fs3: rflash started, upgrade failure will be retried up to 2 times. Please wait...
fs3: Success updating firmware.
```
So there is still a bug, we failed to support below usage:
```
    OpenPOWER BMC specific (using IPMI):
        rflash <noderange> -d <data_directory>
```

Another question, if current firmware version is xxxx, then upgrade to version xxxx, the upgrade process will do again.  Does it make sense?

**Comment by hu-weihua** [comment]
So this issue has been fixed by pr #3973. close the defect.
